component/loki/relabel/relabel.go

```diff
@@ -0,0 +1,214 @@
+type Component struct {
+	mut      sync.RWMutex
+	opts     component.Options
+	rcs      []*relabel.Config
+	receiver loki.LogsReceiver
+	fanout   []loki.LogsReceiver
+
+	entriesProcessed prometheus_client.Counter
+	entriesOutgoing  prometheus_client.Counter
+	cacheHits        prometheus_client.Counter
+	cacheMisses      prometheus_client.Counter
+	cacheSize        prometheus_client.Gauge
+
+	cache *lru.Cache
+}
+
+var (
+	_ component.Component = (*Component)(nil)
+
+	maxCacheSize = 10_000 // TODO(@tpaschalis) Do we make this configurable?
+)
+
+// New creates a new loki.relabel component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	cache, err := lru.New(maxCacheSize)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Component{
+		opts:  o,
+		cache: cache,
+	}
+	c.entriesProcessed = prometheus_client.NewCounter(prometheus_client.CounterOpts{
+		Name: "loki_relabel_entries_processed",
+		Help: "Total number of log entries processed",
+	})
+	c.entriesOutgoing = prometheus_client.NewCounter(prometheus_client.CounterOpts{
+		Name: "loki_relabel_entries_written",
+		Help: "Total number of log entries forwarded",
+	})
+	c.cacheMisses = prometheus_client.NewCounter(prometheus_client.CounterOpts{
+		Name: "loki_relabel_cache_misses",
+		Help: "Total number of cache misses",
+	})
+	c.cacheHits = prometheus_client.NewCounter(prometheus_client.CounterOpts{
+		Name: "loki_relabel_cache_hits",
+		Help: "Total number of cache hits",
+	})
+	c.cacheSize = prometheus_client.NewGauge(prometheus_client.GaugeOpts{
+		Name: "loki_relabel_cache_size",
+		Help: "Total size of relabel cache",
+	})
+
+	for _, metric := range []prometheus_client.Collector{c.entriesProcessed, c.entriesOutgoing, c.cacheMisses, c.cacheHits, c.cacheSize} {
+		err := o.Registerer.Register(metric)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Create and immediately export the receiver which remains the same for
+	// the component's lifetime.
+	c.receiver = make(loki.LogsReceiver)
+	o.OnStateChange(Exports{Receiver: c.receiver})
+
+	// Call to Update() to set the relabelling rules once at the start.
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.receiver:
+			lbls := c.relabel(entry)
+			if len(lbls) == 0 {
+				level.Debug(c.opts.Logger).Log("msg", "dropping entry after relabeling", "labels", entry.Labels.String())
+				continue
+			}
+
+			c.entriesOutgoing.Inc()
+			entry.Labels = lbls
+			for _, f := range c.fanout {
+				select {
+				case <-ctx.Done():
+					return nil
+				case f <- entry:
+				}
+			}
+		}
+	}
+}
+
+func (c *Component) relabel(e loki.Entry) model.LabelSet {
+	c.entriesProcessed.Inc()
+	hash := e.Labels.Fingerprint().String()
+	found, ok := c.cache.Get(hash)
+	if ok {
+		c.cacheHits.Inc()
+		return found.(model.LabelSet)
+	}
```

Comment:
What if there's a hash collision?

Reply:
Unfortunately, then we'd just get wrong relabeling results here.

From what I see, it's the same Fingerprint that Loki uses internally eg. in their ChunkRefs, or the Distributor hashes, taken from Prometheus. It uses the 64-bit variant of FNV-1A. I couldn't find a theoretical collision rate for the algorithm, but [this](https://www.strchr.com/hash_functions#:~:text=Georgi%20%27Sanmayce%27\)%3A-,Core%20i5%20processor,-Wikipedia) source talks about 2M collisions in the set of the words in english wikipedia (~12M words), which is not really an impressive performance...

Our cache sizes shouldn't reach _those_ sizes (we'd probably have more pressing problems, much earlier on), but do we have any real-world feedback of hashing with FNV-1A?

Reply:
The [hashmap code in Prometheus](https://github.com/prometheus/prometheus/blob/0372e259baf014bbade3134fd79bcdfd8cbdef2c/tsdb/agent/series.go#L48-L61) handles collisions by allowing each entry to be a list of label sets. Whenever there's a hashmap lookup, every label set will be compared against the input to check if it matches.

IMO, collisions are unlikely until they happen, and then they're a huge pain to track down. My gut feeling is even if it's a little extra work to implement (and a bit slower on CPU), protecting against collisions with the approach Prometheus uses will probably save us future pain.

Reply:
I've gone ahead with this, and also wrote a test with and without collisions to see that the implementation works as intended.